fortitude/src/rules/obsolescent/deprecated_character_syntax.rs

```diff
@@ -0,0 +1,115 @@
+use ruff_diagnostics::{Diagnostic, Fix, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_source_file::SourceFile;
+use tree_sitter::Node;
+
+/// ## What does it do?
+/// Checks for outdated declarations of `character*N`,
+/// 'character*(*)' and 'character*(expression)'.
+///
+/// ## Why is this bad?
+/// The syntax `character*N` has been replaced by `character(len=N)` in modern
+/// Fortran. Prefer the second form.
+#[derive(ViolationMetadata)]
+pub(crate) struct DeprecatedCharacterSyntax {
+    original: String,
+    dtype: String,
+    length: String,
+}
+
+impl Violation for DeprecatedCharacterSyntax {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Self { original, .. } = self;
+        format!("'{original}' uses deprecated syntax")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        let Self { dtype, length, .. } = self;
+        Some(format!("Replace with '{dtype}(len={length})'"))
+    }
+}
+
+impl AstRule for DeprecatedCharacterSyntax {
+    fn check(
+        _settings: &Settings,
+        node: &Node,
+        source_file: &SourceFile,
+    ) -> Option<Vec<Diagnostic>> {
+        let src = source_file.source_text();
+
+        // Rule only applies to `character`.
+        // Expect child(0) to always be present.
+        let dtype = node.child(0)?;
+        if dtype.kind() != "character" {
+            return None;
+        }
+
+        // If 'kind' field isn't present, exit early
+        let kind = node.child_by_field_name("kind")?;
+        let kind_text = kind.to_text(src)?;
+
+        // If kind does not start with '*', exit early
+        if !kind_text.starts_with('*') {
+            return None;
+        }
+
+        // The '*' should be followed by:
+        // - An integer literal
+        // - '(*)'
+        // - An integer expression within parentheses
+        // To test for the latter, we need to consider two cases. If some
+        // arithmetic takes place, we need to check for a math_expression node
+        // within the kind node. If there isn't, and the kind is instead something
+        // like '(3)', '(N)', or '((((4))))', we need to handle that with regex.
+        let length = match kind
+            .named_descendants()
+            .find(|n| n.kind() == "math_expression")
+        {
+            Some(math_expression) => math_expression.to_text(src)?.to_string(),
+            _ => {
+                // If there is no math_expression node, this may be an integer
+                // literal or '(*)', or it may also be something like '(N)', or
+                // '((((4))))'.
+                // An arbitrary amount of whitespace is permitted.
+                // Strangely, plain '*N' is not allowed.
+                let (_, length, star, expression) = regex_captures!(
+                    r#"^\*(?:\s*(\d+)|\s*\(\s*(\*)\s*\)|\([\s\(]*([[:word:]]+)[\s\)]*\))$"#,
+                    kind_text
+                )?;
```

Comment:
Sort of surprised we have to go for a regex here. Is the tree very complicated at this point?

Reply:
Looking into it, I don't know if we strictly need to. The tree does look quite strange, with `character*(expression)` having an `assumed_size` node even if expression is not `*`. I'll see if I can simplify things.

Reply:
Neat, managed to get it down to just a few lines in https://github.com/PlasmaFAIR/fortitude/pull/354/commits/e101e5bd1e012de0c0426ba3b69167af0b778956.

This isn't at all urgent or necessary, but could the `tree-sitter` grammar be amended so that the first `assumed_size` node is instead something like `character_length_operator`?